src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -644,6 +644,82 @@ protected override void ApplyAnalyzerReferenceRemoved(ProjectId projectId, Analy
+            foreach (var changedDocumentId in changedMappedDocuments)
+            {
+                var oldDoc = projectChanges.OldProject.GetDocument(changedDocumentId)!;
+                var newDoc = projectChanges.NewProject.GetDocument(changedDocumentId)!;
+
+                GetMappedTextChanges(newDoc, oldDoc, mappedChangesBuilder);
+            }
+
+            var mappedChanges = mappedChangesBuilder.ToImmutable();
+            if (mappedChanges.Any())
+            {
+                // Group the mapped text changes by file, then apply all mapped text changes to each file.
+                foreach (var changesForFile in mappedChanges.GroupBy(change => change.fileName))
+                {
+                    using var invisibleEditor = new InvisibleEditor(ServiceProvider.GlobalProvider, changesForFile.Key, GetHierarchy(projectChanges.ProjectId), needsSave: true, needsUndoDisabled: false);
+                    TextEditApplication.UpdateText(changesForFile.Select(change => change.textChange), invisibleEditor.TextBuffer, EditOptions.None);
+                }
+            }
+
+            void GetMappedTextChanges(CodeAnalysis.Document newDocument, CodeAnalysis.Document oldDocument, ArrayBuilder<(string, TextChange)> builder)
+            {
+                var mappingService = newDocument.Services.GetService<ISpanMappingService>();
+                (var mappedSpanResults, var textChanges) = _threadingContext.JoinableTaskFactory.Run<(ImmutableArray<MappedSpanResult>, ImmutableArray<TextChange>)>(async () =>
+                {
+                    var textChanges = (await newDocument.GetTextChangesAsync(oldDocument, CancellationToken.None).ConfigureAwait(false)).ToImmutableArray();
+                    var mappedSpanResults = await mappingService.MapSpansAsync(oldDocument, textChanges.Select(tc => tc.Span), CancellationToken.None).ConfigureAwait(false);
+
+                    return (mappedSpanResults, textChanges);
+                });
+
+                Contract.ThrowIfFalse(mappedSpanResults.Length == textChanges.Length);
+
+                for (var i = 0; i < mappedSpanResults.Length; i++)
+                {
+                    if (ShouldIncludeMappedChange(mappedSpanResults[i], newDocument))
+                    {
+                        builder.Add((mappedSpanResults[i].FilePath, new TextChange(mappedSpanResults[i].Span, textChanges[i].NewText)));
+                    }
+                }
+            }
+
+            static bool ShouldIncludeMappedChange(MappedSpanResult result, CodeAnalysis.Document newDocument)
+            {
+                if (result.IsDefault)
+                {
+                    // There is no appropriate mapping for the span, do not make a change.
+                    return false;
+                }
+
+                var documentIdsForFilePath = newDocument.Project.Solution.GetDocumentIdsWithFilePath(result.FilePath);
```

Comment:
is filepath safe to check?  i'm not really sure what the invariants here.  What if it differs on casing or something?

Reply:
I removed this check and added a different guard before mapping - I originally added this because open aspx files are both in our workspace and have a span mapping service (that points to the exact same location).

However instead of doing this check, I can just check for documents that have a span mapping service and where document.CanApplyChanges() is false.